spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb

```diff
@@ -0,0 +1,321 @@
+# typed: ignore
+
+require 'datadog/tracing/contrib/support/spec_helper'
+require 'rack/test'
+
+require 'securerandom'
+require 'sinatra/base'
+
+begin
+  require 'rack/contrib/json_body_parser'
+rescue LoadError
+  # fallback for old rack-contrib
+  require 'rack/contrib/post_body_content_type_parser'
+end
+
+require 'datadog/tracing'
+require 'datadog/tracing/metadata/ext'
+require 'datadog/tracing/sampling/ext'
+require 'datadog/tracing/contrib/sinatra/ext'
+require 'datadog/tracing/contrib/sinatra/tracer'
+
+require 'datadog/appsec'
+require 'datadog/appsec/contrib/rack/request_middleware'
+
+RSpec.describe 'Sinatra integration tests' do
+  include Rack::Test::Methods
+
+  let(:sorted_spans) do
+    chain = lambda do |start|
+      loop.with_object([start]) do |_, o|
+        # root reached (default)
+        break o if o.last.parent_id == 0
+
+        parent = spans.find { |span| span.span_id == o.last.parent_id }
+
+        # root reached (distributed tracing)
+        break o if parent.nil?
+
+        o << parent
+      end
+    end
+    sort = ->(list) { list.sort_by { |e| chain.call(e).count } }
+    sort.call(spans)
+  end
+
+  let(:sinatra_span) { sorted_spans.reverse.find { |x| x.name == Datadog::Tracing::Contrib::Sinatra::Ext::SPAN_REQUEST } }
+  let(:route_span) { sorted_spans.find { |x| x.name == Datadog::Tracing::Contrib::Sinatra::Ext::SPAN_ROUTE } }
+  let(:rack_span) { sorted_spans.reverse.find { |x| x.name == Datadog::Tracing::Contrib::Rack::Ext::SPAN_REQUEST } }
+
+  let(:appsec_enabled) { true }
+  let(:tracing_enabled) { true }
+
+  before do
+    Datadog.configure do |c|
+      c.tracing.enabled = tracing_enabled
+      c.tracing.instrument :sinatra
+
+      c.appsec.enabled = appsec_enabled
+      c.appsec.instrument :sinatra
+
+      # TODO: test with c.appsec.instrument :rack
+    end
+  end
+
+  around do |example|
+    # Reset before and after each example; don't allow global state to linger.
+    Datadog.registry[:sinatra].reset_configuration!
+    example.run
+    Datadog.registry[:sinatra].reset_configuration!
+  end
```

Comment:
I'm curious about the pattern of `reset_configuration!` usage in all three integration tests: rack does it only after the test, rails never resets, and sinatra does it before AND after the test.

This makes me suspect that some or all of these are either not needed, or missing. Can you doublecheck?

Reply:
I took these from `spec/datadog/tracing/contrib/{rack,sinatra,rails}/integration_test_spec.rb`.

> rails never resets

I think it does (after), see:

- `spec/datadog/tracing/contrib/rails/rails_helper.rb` (especially the big final comment)
- `spec/datadog/tracing/contrib/rails/support/application.rb`

> sinatra does it before

I think this is to eagerly guard against other specs not resetting when they should (which somewhat reduces flakiness, but also makes it less likely to catch the root cause for non-after-resetting specs)